File: src/my_cpp_pkg/src/robot_news_station.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp" //including the interface, same as we used with python sub/pub nodes
//but just a dif syntax compared to python, but serves same purpose^^

using namespace std::chrono_literals;

class RobotNewsStationNode : public rclcpp::Node 
{
public:
    RobotNewsStationNode() : Node("robot_news_station"), robot_name_("R2D2") //node constructor
        // first param is topic name ^ in the double parentheses
    {
        publisher_ = this->create_publisher<example_interfaces::msg::String>("robot_news", 10); // ten is the "queue size"
        timer_ = this->create_wall_timer(0.5s, std::bind(&RobotNewsStationNode::publishNews, this)); //first arg is duration of timer
        RCLCPP_INFO(this->get_logger(), "Robot News Station has been started"); //message to display to indicate construction is complete
    }
 
private:
    void publishNews()
    {
        auto msg = example_interfaces::msg::String();
        msg.data = std::string("Hi, this is ")+ robot_name_ + std::string(" from the robot news station.");
        publisher_->publish(msg); //publishing the message
    }
    std::string robot_name_;
    rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_; //creating publisher attribute and shared pointer
    rclcpp::TimerBase::SharedPtr timer_; //creating a timer to regulate frequency at which publisher publishes messages to topic
};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<RobotNewsStationNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/robot_news_station.py
#!/usr/bin/env python3

#the following node was written using template_py_node.py as a template

import rclpy
from rclpy.node import Node
from example_interfaces.msg import String #importing data type String
#don't forget to add new dependencies, in this case example_interfaces.msg, to package.xml!
 
 
class RobotNewsStationNode(Node): # notice the class attributes defined below in this class declaration!
    def __init__(self): #constructor, methods indicate a publisher and timer will be created upon construction
        super().__init__("robot_news_station") #node name is in the double parentheses ""
        #notice how we made node name the same as the file name, common practice
        #adding additional class methods:
        self.robot_name_ = "C3PO"
        self.publisher_ = self.create_publisher(String, "robot_news", 10) #each topic needs a name and data type --> these are the args here
        #last argument sets buffer of messages sent to 10, can be adjusted but a good baseline
        self.timer_ = self.create_timer(0.5, self.publish_news)
        self.get_logger().info("Robot news station has been started.") #this message will display at end of constructor to indicate in serial monitor construction is complete

    def publish_news(self):
        msg = String()
        msg.data = "Hello, this is " + self.robot_name_ + " from the robot news station." # a message being published by node
        self.publisher_.publish(msg)
 
 
def main(args=None):
    rclpy.init(args=args)
    node = RobotNewsStationNode() #start node
    rclpy.spin(node) #make node spin
    rclpy.shutdown() #shut down node
 
 
if __name__ == "__main__":
    main()
